### PR TITLE
Fetching large chunks

### PR DESCRIPTION
**Pull request by pstibrany**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

This PR fixes issue with fetching chunks that are bigger than 16000 bytes. Under some circumstances it is possible to come across chunks like that, and under some conditions Thanos may currently fail to fetch such chunks. This PR adds fetching of remaining chunk data from the bucket.

## Verification

Added unit test. Also tested manually against real block found in our test cluster.

Fixes issue https://github.com/thanos-io/thanos/issues/2945.

### CODE REVIEW

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -2050,11 +2058,44 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, offs []uint32, seq i
...
+		if len(cb) >= chLen {
+			r.chunks[chunkRef] = rawChunk(cb[n:chLen])
+			continue
 		}
-		cid := uint64(seq<<32) | uint64(o)
-		r.chunks[cid] = rawChunk(cb[n : n+int(l)+1])
+
+		// If we didn't fetch enough data for the chunk, fetch more. This can only really happen for last
+		// chunk in the list of fetched chunks, otherwise partitioner would merge fetch ranges together.
+		r.mtx.Unlock()
```
Comment (bwplotka) [inline pkg/store/bucket.go:2074]: Worth to mention what this mutex is saving from. Probably changing name would be nice. I know it was like this before, but maybe we can improve that part. I assume it's for stats, right?

Comment (pstibrany) [reply pkg/store/bucket.go:2074]: It's both for stats and storing chunks. I will document and rename it.

Comment (pstibrany) [reply pkg/store/bucket.go:2074]: Also `chunkBytes`.

**Comment by bwplotka** [comment]
Thanks!

**Comment by pracucci** [comment]
I'm joining the party very late, but good job @pstibrany ! 👏

**Pull request merged**